lib/linked_list.rb

```diff
@@ -12,152 +12,330 @@ def initialize(value, next_node = nil)
-    # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity:  
-    # Space Complexity
-    def reverse
-      raise NotImplementedError
+  # method to delete the first node found with specified value
+  # Time Complexity: O(n) where n is the length of the linke list
+  # Space Complexity: O(1)
+  def delete(value)
+    return if @head == nil
+
+    current = @head
+    last = nil
+
+    until current == nil
+      if value == current.data
+        if @head.data == value
+          @head = current.next
+        else
+          last.next = last.next.next
+        end
+      end
     end
 
+    last = current
+    current = current.next
+  end
+
+  # return if !@head
+
+  # previous = nil
+  # current = @head
+
+  # until !current
+  #   if current.data == value
+  #     @head.data == value ? @head = current.next : previous.next = current.next
+  #   end
+
+  #   previous = current
+  #   current = current.next
+  # end
+  # end
 
-    ## Advanced Exercises
-    # returns the value at the middle element in the singly linked list
-    # Time Complexity:  
-    # Space Complexity
-    def find_middle_value
-      raise NotImplementedError
+  # method to reverse the singly linked list
+  # note: the nodes should be moved and not just the values in the nodes
+  # Time Complexity: O(n) where n is the length of the linked list
+  # Space Complexity: O(1)
+  def reverse
+    current = @head
+    last_node = nil
+    next_node = current.next
+
+    if current == nil
+      return nil
+    end
+
+    while current.next != nil
+      current.next = last_node
+      last_node = current
+      current = next_node
+      next_node = current.next
     end
 
-    # find the nth node from the end and return its value
-    # assume indexing starts at 0 while counting to n
-    # Time Complexity:  
-    # Space Complexity
-    def find_nth_from_end(n)
-      raise NotImplementedError
+    current.next = last_node
+    @head = current
+    return @head.data
+  end
+
+  ## Advanced Exercises
+  # returns the value at the middle element in the singly linked list
+  # Time Complexity: O(n) where n is the length of the linked list
+  # Space Complexity: O(1)
+  def find_middle_value
+    slow = @head
+    fast = @head
+
+    if @head == nil
+      return nil
     end
 
-    # checks if the linked list has a cycle. A cycle exists if any node in the
-    # linked list links to a node already visited.
-    # returns true if a cycle is found, false otherwise.
-    # Time Complexity:  
-    # Space Complexity
-    def has_cycle
-      raise NotImplementedError
+    while fast != nil && fast.next != nil
+      fast = fast.next.next
+      slow = slow.next
     end
+    return slow
+  end
+
+  # find the nth node from the end and return its value
+  # assume indexing starts at 0 while counting to n
+  # Time Complexity: O(n) where n is the length of the linked list
+  # Space Complexity: O(1)
+  def find_nth_from_end(n)

```

Comment:
Could you do this one in a similar manner to `has_cycle` so that you don't have to find the length before finding the nth from the end?